cmake/multi_image.cmake

```diff
@@ -125,6 +125,82 @@ function(add_child_image_from_source)
+  # | - prj_<desc>.conf (B)
+  # | - boards DIRECTORY
+  # | | - <board>.conf (C)
+  # | | - <board>_<desc>.conf (D)
+  # | - child_image DIRECTORY
+  # | | <ACI_NAME> DIRECTORY
+  # | | - boards DIRECTORY
+  # | | | - <board>.conf (E)        If present, use instead of (C)
+  # | | | - <board>_<desc>.conf (F) If present, use instead of (D)
+  # | | - prj.conf (G)              If present, use instead of (A)
+  # | |                             Note that (C) is ignored if this is present.
+  # | |                             Use (E) instead.
+  # | | - prj_<desc>.conf (H)       If present, used instead of (B) when user
+  # | |                             specify `-DCONF_FILE=prj_<desc>.conf for
+  # | |                             parent image. Note that any (C) is ignored
+  # | |                             if this is present. Use (F) instead.
+  # | | - <ACI_name>_prj.conf (I)   Fragment, used together with (A) and (C)
+  # | | - <ACI_name>_prj_<desc>.conf (J) Fragment, used together with (B) and (D)
+
+  set(alt_conf ${APPLICATION_SOURCE_DIR}/child_image/${ACI_NAME})
+  if (EXISTS ${alt_conf})
+    # First of all we need to clean up the 'CONF_FILE' variable. It could be
+    # a list if (C) or (D) is present. If (C) or (D) is present and
+    # (G) or (H) is present (C) or (D)  will not be used. Hence for the purpose
+    # of these checks, we remove the (C) or (D) entry from  the CONF_FILE
+    # variable if it exists.
+    list(LENGTH CONF_FILE CONF_FILE_LENGTH)
+    if (NOT ${CONF_FILE_LENGTH} EQUAL 1)
+      list(GET CONF_FILE_AS_LIST 0 safe_conf_file)
+    else()
+      set(safe_conf_file ${CONF_FILE})
+    endif()
+
+    # Now that we have a conf file variable which we know only contain a single
+    # entry, identify if any special configuration variant is used.
+    # These are the prj_<alternative name>.conf variants.
+    if (${safe_conf_file} MATCHES "prj_(.*).conf" AND
+         (NOT ${CMAKE_MATCH_1} STREQUAL ${BOARD}))
+      set(prj_conf_file_name prj_${CMAKE_MATCH_1}.conf)
+      set(board_conf_file_name ${ACI_BOARD}_${CMAKE_MATCH_1}.conf)
+    else()
+      set(prj_conf_file_name prj.conf)
+      set(board_conf_file_name ${ACI_BOARD}.conf)
+    endif()
+
+    # Check for CONF_FILE. This is the base configuration, as opposed
+    # to a fragment which is appended to the configuration.
+    # If found, pass it as CONF_FILE to the child image.
+    set(child_image_conf_file ${alt_conf}/${prj_conf_file_name})
+    if (EXISTS ${child_image_conf_file})
+      set(${ACI_NAME}_CONF_FILE ${child_image_conf_file})
+    endif()
+
+    # Check for configuration fragment. The contents of these are appended
+    # to the project configuration, as opposed to the CONF_FILE which is used
+    # as the base configuration.
+    set(child_image_conf_fragment ${alt_conf}/${ACI_NAME}_${prj_conf_file_name})
+    if (EXISTS ${child_image_conf_fragment})
+      add_overlay_config(${ACI_NAME} ${child_image_conf_fragment})
+    endif()
```

Comment:
I know this is what I initially proposed here:
https://github.com/nrfconnect/sdk-nrf/pull/3247/#discussion_r514322303

but maybe we should consider moving those one level up, like this:
```
<sample>
  |--- prj.conf
  |--- prj_<desc>.conf
  |--- <ACI_name>_prj.conf         <-- Fragment, use together with sample/<ACI_name>/prj.conf
  |--- <ACI_name>_prj_<desc>.conf  <-- Fragment, use together with sample/<ACI_name>/prj_<desc>.conf
  |--- child_image
         |--- <ACI_name>  <-- This is a folder
                |--- prj.conf         <-- If present, use instead of sample/<ACI_name>/prj.conf
                |--- prj_<desc>.conf  <-- If present, used when user specify `-DCONF_FILE=prj_<desc>.conf for parent image instead of sample/<ACI_name>/prj_<desc>.conf
```
that might make it clearer that everything placed directly under `<sample>` is used together with the `<ACI_name>` normal configs.
And everything placed under `<sample>/child_image/<ACI_name>` will be used instead of the `<ACI_name>` normal configs.

@hakonfam what do you think ? What will make it easier for user to understand.

Reply:
TBH I don't think your latest proposal is easier to understand than the original one, since it is still not intuitive, and you would have to read some documentation to figure out which is which. (Why does a new folder imply that it is not a fragment?)

The only way I can imagine making this a bit more intuitive is if we add "fragment" somewhere in the name.

```<sample>
  |--- prj.conf
  |--- prj_<desc>.conf
  |--- child_image
         |--- <ACI_name>  <-- This is a folder
                |--- fragment.conf  <-- Fragment, use together with sample/<ACI_name>/prj.conf
                |--- fragment_<desc>.conf <-- Fragment, use together with sample/<ACI_name>/prj_<desc>.conf
                |--- prj.conf         <-- If present, use instead of sample/<ACI_name>/prj.conf
                |--- prj_<desc>.conf  <-- If present, used when user specify `-DCONF_FILE=prj_<desc>.conf for parent image instead of sample/<ACI_name>/prj_<desc>.conf
```

Reply:
> TBH I don't think your latest proposal is easier to understand than the original one, since it is still not intuitive, and you would have to read some documentation to figure out which is which.

Of course, users will need to read the docs.

My main concern is that there are users who will mess up the:
`child_image/<ACI>/<ACI>_prj.conf` vs `child_image/<ACI>/prj.conf` name.

> (Why does a new folder imply that it is not a fragment?)

In itself, it doesn't, but users only needing a `ACI` prj.conf fragment together with the samples normal conf file doesn't need to create a `child_image/<ACI>` folder.
This minimizes the risk that a user mistakenly creates a: `child_image/<ACI>/<ACI>_prj.conf` file, when he meant to create a `child_image/<ACI>/prj.conf`file.

Secondly, we can print a warning, if he creates a `child_image/<ACI>/<ACI>_prj.conf`, like:
```
WARNING child_image/<ACI>/<ACI>_prj.conf` file discovered. `<ACI>_prj.conf` should be at application root, ${APP_DIR}.
If this is a full conf file for `<ACI>`, then it must be named `prj.conf` instead.
```

When both kind of files are located in same folder, it will be easier for a user to make such a mistake.

Of course, a user may still create a `<ACI>_prj.conf` when he actually wants a `child_image/<ACI>/prj.conf`.
But in that case, the `<ACI>_prj.conf` must be expected to be a full config file, thus chances are that he is anyway setting most config settings, and settings in this file will have precedence over the `<ACI>` default conf file, so this may not have much practical impact.

> The only way I can imagine making this a bit more intuitive is if we add "fragment" somewhere in the name.

I'm not sure users will find it easier.
All `*.conf` files are Kconfig fragments, even the `prj.conf`

Reply:
Some additional inspiration where I would like your feedback.

A `<board>.conf` is used for a Kconfig fragment that should be used together with a `prj.conf`.

So mayby we should use `<ACI>.conf` for Kconfig fragments that is to be used together with a `prj.conf`, and then reserve the `prj.conf` for the case where it should be used instead of the sample's default `prj.conf`.
That will give us `<ACI>.conf` and `prj.conf` and minimize the risk of misunderstanding `<ACI>_prj.conf` vs `prj.conf`.

So we can have:
```
|--- prj.conf
  |--- prj_<desc>.conf
  |--- child_image
         |--- <ACI_name>  <-- This is a folder
                |--- <ACI_name>.conf  <-- Fragment, use together with prj.conf
                |--- <ACI_name>_<desc>.conf <-- Fragment, use together with prj_<desc>.conf
                |--- prj.conf         <-- If present, use instead of sample/<ACI_name>/prj.conf
                |--- prj_<desc>.conf  <-- If present, used when user specify `-DCONF_FILE=prj_<desc>.conf for parent image instead of sample/<ACI_name>/prj_<desc>.conf
```

We could then always apply a `<ACI_name>.conf` to the `prj.conf` regardless of whether that `prj.conf` resides in
1. `sample/<ACI_name>/prj.conf`
2. `child_image/<ACI_name>/prj.conf`

although it doesn't makes little sense to have `child_image/<ACI_name>/prj.conf` together with `child_image/<ACI_name>/<ACI_name>.conf` (situation 2.), then there is no harm in doing so, and it will make the logic for when `child_image/<ACI_name>/<ACI_name>.conf` is merged consistent, as it will always be merged with a `prj.conf`.